### PR TITLE
Add zip functions in vector module to iterate over 2 vectors concurrently

### PR DESCRIPTION
**Pull request by movekevin**
### Description
This is a common operations for iterating over vector inputs.

### Test Plan
New unit tests

### CODE REVIEW

File: aptos-move/framework/move-stdlib/sources/vector.move [hunk 1]
```diff
@@ -263,6 +266,51 @@ module std::vector {
         }
     }
 
+    /// Apply the function to each pair of elements in the two given vectors, consuming them.
+    public inline fun zip<Element1, Element2>(v1: vector<Element1>, v2: vector<Element2>, f: |Element1, Element2|) {
```
Comment (0xbe1) [inline aptos-move/framework/move-stdlib/sources/vector.move:270]: can you also add a version that instead returns `zip<Element1, Element2, A>(v1: vector<Element1>, v2: vector<Element2>, f: |Element1, Element2| A)`

Comment (movekevin) [reply aptos-move/framework/move-stdlib/sources/vector.move:270]: Added zip_map and zip_map_ref

**Review by xindingw — APPROVED** [review]

File: aptos-move/framework/move-stdlib/sources/vector.move [hunk 2]
```diff
@@ -263,6 +266,51 @@ module std::vector {
...
+
+    /// Apply the function to each pair of elements in the two given vectors in the reverse order, consuming them.
+    /// This errors out if the vectors are not of the same length.
+    public inline fun zip_reverse<Element1, Element2>(
+        v1: vector<Element1>,
+        v2: vector<Element2>,
+        f: |Element1, Element2|,
+    ) {
+        let len = length(&v1);
+        // We can't use the constant EVECTORS_LENGTH_MISMATCH here as all calling code would then need to define it
```
Comment (lightmark) [inline aptos-move/framework/move-stdlib/sources/vector.move:285]: when could this be solved... This is ugly. Can we make this not inline and later convert it to inline if we solved the scope issue of inline function?

**Review by lightmark — APPROVED** [review]

File: aptos-move/framework/move-stdlib/sources/vector.move (hunk 2 above)
Comment (movekevin) [reply aptos-move/framework/move-stdlib/sources/vector.move:285]: There's a Slack thread on this where we're discussing :)

Comment (brmataptos) [reply aptos-move/framework/move-stdlib/sources/vector.move:285]: We need a way to refer to this module's const in the place where it is inlined.  But I suppose that's a bytecode issue.

**Pull request merged**